File: csharp/Svix/IngestSource.cs
// this file is @generated
#nullable enable
using Microsoft.Extensions.Logging;
using Svix.Models;

namespace Svix
{
    public class IngestSourceListOptions : SvixOptionsBase
    {
        public ulong? Limit { get; set; }
        public string? Iterator { get; set; }
        public Ordering? Order { get; set; }

        public new Dictionary<string, string> QueryParams()
        {
            return SerializeParams(
                new Dictionary<string, object?>
                {
                    { "limit", Limit },
                    { "iterator", Iterator },
                    { "order", Order },
                }
            );
        }
    }

    public class IngestSourceCreateOptions : SvixOptionsBase
    {
        public string? IdempotencyKey { get; set; }

        public new Dictionary<string, string> HeaderParams()
        {
            return SerializeParams(
                new Dictionary<string, object?> { { "idempotency-key", IdempotencyKey } }
            );
        }
    }

    public class IngestSourceRotateTokenOptions : SvixOptionsBase
    {
        public string? IdempotencyKey { get; set; }

        public new Dictionary<string, string> HeaderParams()
        {
            return SerializeParams(
                new Dictionary<string, object?> { { "idempotency-key", IdempotencyKey } }
            );
        }
    }

    public class IngestSource(SvixClient client)
    {
        readonly SvixClient _client = client;

        /// <summary>
        /// List of all the organization's Ingest Sources.
        /// </summary>
        public async Task<ListResponseIngestSourceOut> ListAsync(
            IngestSourceListOptions? options = null,
            CancellationToken cancellationToken = default
        )
        {
            try
            {
                var response =
                    await _client.SvixHttpClient.SendRequestAsync<ListResponseIngestSourceOut>(
                        method: HttpMethod.Get,
                        path: "/ingest/api/v1/source",
                        queryParams: options?.QueryParams(),
                        headerParams: options?.HeaderParams(),
                        cancellationToken: cancellationToken
                    );
                return response.Data;
            }
            catch (ApiException e)
            {
                _client.Logger?.LogError(e, $"{nameof(ListAsync)} failed");

                throw;
            }
        }

        /// <summary>
        /// List of all the organization's Ingest Sources.
        /// </summary>
        public ListResponseIngestSourceOut List(IngestSourceListOptions? options = null)
        {
            try
            {
                var response = _client.SvixHttpClient.SendRequest<ListResponseIngestSourceOut>(
                    method: HttpMethod.Get,
                    path: "/ingest/api/v1/source",
                    queryParams: options?.QueryParams(),
                    headerParams: options?.HeaderParams()
                );
                return response.Data;
            }
            catch (ApiException e)
            {
                _client.Logger?.LogError(e, $"{nameof(List)} failed");

                throw;
            }
        }

        /// <summary>
        /// Create Ingest Source.
        /// </summary>
        public async Task<IngestSourceOut> CreateAsync(
            IngestSourceIn ingestSourceIn,
            IngestSourceCreateOptions? options = null,
            CancellationToken cancellationToken = default
        )
        {
            ingestSourceIn =
                ingestSourceIn ?? throw new ArgumentNullException(nameof(ingestSourceIn));
            try
            {
                var response = await _client.SvixHttpClient.SendRequestAsync<IngestSourceOut>(
                    method: HttpMethod.Post,
                    path: "/ingest/api/v1/source",
                    queryParams: options?.QueryParams(),
                    headerParams: options?.HeaderParams(),
                    content: ingestSourceIn,
                    cancellationToken: cancellationToken
                );
                return response.Data;
            }
            catch (ApiException e)
            {
                _client.Logger?.LogError(e, $"{nameof(CreateAsync)} failed");

                throw;
            }
        }

        /// <summary>
        /// Create Ingest Source.
        /// </summary>
        public IngestSourceOut Create(
            IngestSourceIn ingestSourceIn,
            IngestSourceCreateOptions? options = null
        )
        {
            ingestSourceIn =
                ingestSourceIn ?? throw new ArgumentNullException(nameof(ingestSourceIn));
            try
            {
                var response = _client.SvixHttpClient.SendRequest<IngestSourceOut>(
                    method: HttpMethod.Post,
                    path: "/ingest/api/v1/source",
                    queryParams: options?.QueryParams(),
                    headerParams: options?.HeaderParams(),
                    content: ingestSourceIn
                );
                return response.Data;
            }
            catch (ApiException e)
            {
                _client.Logger?.LogError(e, $"{nameof(Create)} failed");

                throw;
            }
        }

        /// <summary>
        /// Get an Ingest Source by id or uid.
        /// </summary>
        public async Task<IngestSourceOut> GetAsync(
            string sourceId,
            CancellationToken cancellationToken = default
        )
        {
            try
            {
                var response = await _client.SvixHttpClient.SendRequestAsync<IngestSourceOut>(
                    method: HttpMethod.Get,
                    path: "/ingest/api/v1/source/{source_id}",
                    pathParams: new Dictionary<string, string> { { "source_id", sourceId } },
                    cancellationToken: cancellationToken
                );
                return response.Data;
            }
            catch (ApiException e)
            {
                _client.Logger?.LogError(e, $"{nameof(GetAsync)} failed");

                throw;
            }
        }

        /// <summary>
        /// Get an Ingest Source by id or uid.
        /// </summary>
        public IngestSourceOut Get(string sourceId)
        {
            try
            {
                var response = _client.SvixHttpClient.SendRequest<IngestSourceOut>(
                    method: HttpMethod.Get,
                    path: "/ingest/api/v1/source/{source_id}",
                    pathParams: new Dictionary<string, string> { { "source_id", sourceId } }
                );
                return response.Data;
            }
            catch (ApiException e)
            {
                _client.Logger?.LogError(e, $"{nameof(Get)} failed");

                throw;
            }
        }

        /// <summary>
        /// Update an Ingest Source.
        /// </summary>
        public async Task<IngestSourceOut> UpdateAsync(
            string sourceId,
            IngestSourceIn ingestSourceIn,
            CancellationToken cancellationToken = default
        )
        {
            ingestSourceIn =
                ingestSourceIn ?? throw new ArgumentNullException(nameof(ingestSourceIn));
            try
            {
                var response = await _client.SvixHttpClient.SendRequestAsync<IngestSourceOut>(
                    method: HttpMethod.Put,
                    path: "/ingest/api/v1/source/{source_id}",
                    pathParams: new Dictionary<string, string> { { "source_id", sourceId } },
                    content: ingestSourceIn,
                    cancellationToken: cancellationToken
                );
                return response.Data;
            }
            catch (ApiException e)
            {
                _client.Logger?.LogError(e, $"{nameof(UpdateAsync)} failed");

                throw;
            }
        }

        /// <summary>
        /// Update an Ingest Source.
        /// </summary>
        public IngestSourceOut Update(string sourceId, IngestSourceIn ingestSourceIn)
        {
            ingestSourceIn =
                ingestSourceIn ?? throw new ArgumentNullException(nameof(ingestSourceIn));
            try
            {
                var response = _client.SvixHttpClient.SendRequest<IngestSourceOut>(
                    method: HttpMethod.Put,
                    path: "/ingest/api/v1/source/{source_id}",
                    pathParams: new Dictionary<string, string> { { "source_id", sourceId } },
                    content: ingestSourceIn
                );
                return response.Data;
            }
            catch (ApiException e)
            {
                _client.Logger?.LogError(e, $"{nameof(Update)} failed");

                throw;
            }
        }

        /// <summary>
        /// Delete an Ingest Source.
        /// </summary>
        public async Task<bool> DeleteAsync(
            string sourceId,
            CancellationToken cancellationToken = default
        )
        {
            try
            {
                var response = await _client.SvixHttpClient.SendRequestAsync<bool>(
                    method: HttpMethod.Delete,
                    path: "/ingest/api/v1/source/{source_id}",
                    pathParams: new Dictionary<string, string> { { "source_id", sourceId } },
                    cancellationToken: cancellationToken
                );
                return response.Data;
            }
            catch (ApiException e)
            {
                _client.Logger?.LogError(e, $"{nameof(DeleteAsync)} failed");

                throw;
            }
        }

        /// <summary>
        /// Delete an Ingest Source.
        /// </summary>
        public bool Delete(string sourceId)
        {
            try
            {
                var response = _client.SvixHttpClient.SendRequest<bool>(
                    method: HttpMethod.Delete,
                    path: "/ingest/api/v1/source/{source_id}",
                    pathParams: new Dictionary<string, string> { { "source_id", sourceId } }
                );
                return response.Data;
            }
            catch (ApiException e)
            {
                _client.Logger?.LogError(e, $"{nameof(Delete)} failed");

                throw;
            }
        }

        /// <summary>
        /// Rotate the Ingest Source's Url Token.
        ///
        /// This will rotate the ingest source's token, which is used to
        /// construct the unique `ingestUrl` for the source. Previous tokens
        /// will remain valid for 48 hours after rotation. The token can be
        /// rotated a maximum of three times within the 48-hour period.
        /// </summary>
        public async Task<RotateTokenOut> RotateTokenAsync(
            string sourceId,
            IngestSourceRotateTokenOptions? options = null,
            CancellationToken cancellationToken = default
        )
        {
            try
            {
                var response = await _client.SvixHttpClient.SendRequestAsync<RotateTokenOut>(
                    method: HttpMethod.Post,
                    path: "/ingest/api/v1/source/{source_id}/token/rotate",
                    pathParams: new Dictionary<string, string> { { "source_id", sourceId } },
                    queryParams: options?.QueryParams(),
                    headerParams: options?.HeaderParams(),
                    cancellationToken: cancellationToken
                );
                return response.Data;
            }
            catch (ApiException e)
            {
                _client.Logger?.LogError(e, $"{nameof(RotateTokenAsync)} failed");

                throw;
            }
        }

        /// <summary>
        /// Rotate the Ingest Source's Url Token.
        ///
        /// This will rotate the ingest source's token, which is used to
        /// construct the unique `ingestUrl` for the source. Previous tokens
        /// will remain valid for 48 hours after rotation. The token can be
        /// rotated a maximum of three times within the 48-hour period.
        /// </summary>
        public RotateTokenOut RotateToken(
            string sourceId,
            IngestSourceRotateTokenOptions? options = null
        )
        {
            try
            {
                var response = _client.SvixHttpClient.SendRequest<RotateTokenOut>(
                    method: HttpMethod.Post,
                    path: "/ingest/api/v1/source/{source_id}/token/rotate",
                    pathParams: new Dictionary<string, string> { { "source_id", sourceId } },
                    queryParams: options?.QueryParams(),
                    headerParams: options?.HeaderParams()
                );
                return response.Data;
            }
            catch (ApiException e)
            {
                _client.Logger?.LogError(e, $"{nameof(RotateToken)} failed");

                throw;
            }
        }
    }
}

File: svix-cli/src/cmds/api/ingest_source.rs
use clap::{Args, Subcommand};
use svix::api::*;

use crate::json::JsonOf;

#[derive(Args, Clone)]
pub struct IngestSourceListOptions {
    /// Limit the number of returned items
    #[arg(long)]
    pub limit: Option<i32>,
    /// The iterator returned from a prior invocation
    #[arg(long)]
    pub iterator: Option<String>,
    /// The sorting order of the returned items
    #[arg(long)]
    pub order: Option<Ordering>,
}

impl From<IngestSourceListOptions> for svix::api::IngestSourceListOptions {
    fn from(value: IngestSourceListOptions) -> Self {
        let IngestSourceListOptions {
            limit,
            iterator,
            order,
        } = value;
        Self {
            limit,
            iterator,
            order,
        }
    }
}

#[derive(Args, Clone)]
pub struct IngestSourceCreateOptions {
    #[arg(long)]
    pub idempotency_key: Option<String>,
}

impl From<IngestSourceCreateOptions> for svix::api::IngestSourceCreateOptions {
    fn from(value: IngestSourceCreateOptions) -> Self {
        let IngestSourceCreateOptions { idempotency_key } = value;
        Self { idempotency_key }
    }
}

#[derive(Args, Clone)]
pub struct IngestSourceRotateTokenOptions {
    #[arg(long)]
    pub idempotency_key: Option<String>,
}

impl From<IngestSourceRotateTokenOptions> for svix::api::IngestSourceRotateTokenOptions {
    fn from(value: IngestSourceRotateTokenOptions) -> Self {
        let IngestSourceRotateTokenOptions { idempotency_key } = value;
        Self { idempotency_key }
    }
}

#[derive(Args)]
#[command(args_conflicts_with_subcommands = true, flatten_help = true)]
pub struct IngestSourceArgs {
    #[command(subcommand)]
    pub command: IngestSourceCommands,
}

#[derive(Subcommand)]
pub enum IngestSourceCommands {
    /// List of all the organization's Ingest Sources.
    List {
        #[clap(flatten)]
        options: IngestSourceListOptions,
    },
    /// Create Ingest Source.
    Create {
        ingest_source_in: JsonOf<IngestSourceIn>,
        #[clap(flatten)]
        options: IngestSourceCreateOptions,
    },
    /// Get an Ingest Source by id or uid.
    Get { source_id: String },
    /// Update an Ingest Source.
    Update {
        source_id: String,
        ingest_source_in: JsonOf<IngestSourceIn>,
    },
    /// Delete an Ingest Source.
    Delete { source_id: String },
    /// Rotate the Ingest Source's Url Token.
    ///
    /// This will rotate the ingest source's token, which is used to
    /// construct the unique `ingestUrl` for the source. Previous tokens
    /// will remain valid for 48 hours after rotation. The token can be
    /// rotated a maximum of three times within the 48-hour period.
    RotateToken {
        source_id: String,
        #[clap(flatten)]
        options: IngestSourceRotateTokenOptions,
    },
}

impl IngestSourceCommands {
    pub async fn exec(
        self,
        client: &Svix,
        color_mode: colored_json::ColorMode,
    ) -> anyhow::Result<()> {
        match self {
            Self::List { options } => {
                let resp = client.ingest().source().list(Some(options.into())).await?;
                crate::json::print_json_output(&resp, color_mode)?;
            }
            Self::Create {
                ingest_source_in,
                options,
            } => {
                let resp = client
                    .ingest()
                    .source()
                    .create(ingest_source_in.into_inner(), Some(options.into()))
                    .await?;
                crate::json::print_json_output(&resp, color_mode)?;
            }
            Self::Get { source_id } => {
                let resp = client.ingest().source().get(source_id).await?;
                crate::json::print_json_output(&resp, color_mode)?;
            }
            Self::Update {
                source_id,
                ingest_source_in,
            } => {
                let resp = client
                    .ingest()
                    .source()
                    .update(source_id, ingest_source_in.into_inner())
                    .await?;
                crate::json::print_json_output(&resp, color_mode)?;
            }
            Self::Delete { source_id } => {
                client.ingest().source().delete(source_id).await?;
            }
            Self::RotateToken { source_id, options } => {
                let resp = client
                    .ingest()
                    .source()
                    .rotate_token(source_id, Some(options.into()))
                    .await?;
                crate::json::print_json_output(&resp, color_mode)?;
            }
        }

        Ok(())
    }
}
